lib/Doctrine/MongoDB/Aggregation/Builder.php

```diff
@@ -0,0 +1,262 @@
+    {
+        return $this->addStage(new Stage\Limit($this, $limit));
+    }
+
+    /**
+     * Filters the documents to pass only the documents that match the specified condition(s) to the next pipeline stage.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/match/
+     *
+     * @return Stage\Match
+     */
+    public function match()
+    {
+        return $this->addStage(new Stage\Match($this));
+    }
+
+    /**
+     * Takes the documents returned by the aggregation pipeline and writes them to a specified collection. This must be the last stage in the pipeline.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/out/
+     *
+     * @param string $collection
+     * @return Stage\Out
+     */
+    public function out($collection)
+    {
+        return $this->addStage(new Stage\Out($this, $collection));
+    }
+
+    /**
+     * Passes along the documents with only the specified fields to the next stage in the pipeline. The specified fields can be existing fields from the input documents or newly computed fields.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/project/
+     *
+     * @return Stage\Project
+     */
+    public function project()
+    {
+        return $this->addStage(new Stage\Project($this));
+    }
+
+    /**
+     * Returns a query expression to be used in match stages
+     *
+     * @return QueryExpr
+     */
+    public function matchExpr()
+    {
+        return new QueryExpr();
+    }
+
+    /**
+     * Restricts the contents of the documents based on information stored in the documents themselves.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/redact/
+     *
+     * @return Stage\Redact
+     */
+    public function redact()
+    {
+        return $this->addStage(new Stage\Redact($this));
+    }
+
+    /**
+     * Skips over the specified number of documents that pass into the stage and passes the remaining documents to the next stage in the pipeline.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/skip/
+     *
+     * @param integer $skip
+     * @return Stage\Skip
+     */
+    public function skip($skip)
+    {
+        return $this->addStage(new Stage\Skip($this, $skip));
+    }
+
+    /**
+     * Sorts all input documents and returns them to the pipeline in sorted order.
+     *
+     * If sorting by multiple fields, the first argument should be an array of
+     * field name (key) and order (value) pairs.
+     *
+     * @see http://docs.mongodb.org/manual/reference/operator/aggregation/sort/
+     *
+     * @param array|string $fieldName Field name or array of field/order pairs
+     * @param integer|string $order   Field order (if one field is specified)
+     * @return Stage\Sort
+     */
+    public function sort($fieldName, $order = null)
+    {
+        return $this->addStage(new Stage\Sort($this, $fieldName, $order));
```

Comment:
Didn't know MongoDB optimizes this. The opportunity presented itself by overwriting the ```sort()``` method within the ```Sort``` stage, that's where this optimization happens. I can refactor though to leave it out - this would create two sort and limit stages though.